docs/SAFETY-MACROS.md

```diff
@@ -0,0 +1,432 @@
+
+[//]: # (DO NOT DIRECTLY MODIFY THIS FILE:)
+[//]: # (The code in this file is generated from scripts/s2n_safety_macros.py and any modifications)
+[//]: # (should be in there.)
+
+# S2N Safety Macros
+
+## Macros for functions that return `S2N_RESULT`
+
+
+### RESULT_BAIL(error)
+
+Sets the global `s2n_errno` to `error` and returns with an `S2N_RESULT_ERROR`
+
+
+### RESULT_ENSURE(condition, error)
+
+Ensures the `condition` is `true`, otherwise the function will `RESULT_BAIL` with `error`
+
+
+### RESULT_DEBUG_ENSURE(condition, error)
+
+Ensures the `condition` is `true`, otherwise the function will `RESULT_BAIL` with `error`
+
+NOTE: The condition will _only_ be checked when the code is compiled in debug mode.
+      In release mode, the check is removed.
+
+
+### RESULT_ENSURE_OK(result, error)
+
+Ensures `s2n_result_is_ok(result)`, otherwise the function will `RESULT_BAIL` with `error`
+
+This can be useful for overriding the global `s2n_errno`
+
+
+### RESULT_ENSURE_GTE(a, b)
+
+Ensures `a` is greater than or equal to `b`, otherwise the function will `RESULT_BAIL` with a `S2N_ERR_SAFETY` error
+
+
+### RESULT_ENSURE_LTE(a, b)
+
+Ensures `a` is less than or equal to `b`, otherwise the function will `RESULT_BAIL` with a `S2N_ERR_SAFETY` error
+
+
+### RESULT_ENSURE_GT(a, b)
+
+Ensures `a` is greater than `b`, otherwise the function will `RESULT_BAIL` with a `S2N_ERR_SAFETY` error
+
+
+### RESULT_ENSURE_LT(a, b)
+
+Ensures `a` is less than `b`, otherwise the function will `RESULT_BAIL` with a `S2N_ERR_SAFETY` error
+
+
+### RESULT_ENSURE_EQ(a, b)
+
+Ensures `a` is equal to `b`, otherwise the function will `RESULT_BAIL` with a `S2N_ERR_SAFETY` error
+
+
+### RESULT_ENSURE_NE(a, b)
+
+Ensures `a` is not equal to `b`, otherwise the function will `RESULT_BAIL` with a `S2N_ERR_SAFETY` error
+
+
+### RESULT_ENSURE_INCLUSIVE_RANGE(min, n, max)
+
+Ensures `min <= n <= max`, otherwise the function will `RESULT_BAIL` with `S2N_ERR_SAFETY`
+
+
+### RESULT_ENSURE_EXCLUSIVE_RANGE(min, n, max)
+
+Ensures `min < n < max`, otherwise the function will `RESULT_BAIL` with `S2N_ERR_SAFETY`
+
+
+### RESULT_ENSURE_REF(x)
+
+Ensures `x` is a readable reference, otherwise the function will `RESULT_BAIL` with `S2N_ERR_NULL`
+
+
+### RESULT_ENSURE_MUT(x)
+
+Ensures `x` is a mutable reference, otherwise the function will `RESULT_BAIL` with `S2N_ERR_NULL`
+
+
+### RESULT_PRECONDITION(result)
+
+Ensures the `result` is `S2N_RESULT_OK`, otherwise the function will return an error signal
+
+
+### RESULT_POSTCONDITION(result)
+
+Ensures the `result` is `S2N_RESULT_OK`, otherwise the function will return an error signal
+
```

Comment:
Does this sufficiently describe what these are for? We might want to mention that this boils down to __builtin_expect

Reply:
I can copy the description from `DEBUG_ENSURE` for POSTCONDITION